Order interposed sychronizations

mmusgrov opened the pull request:
Fixes #18737

The fix is to reorder interposed synchronizations such that Agroal runs last.
The new order is [OTHER, HIBERNATE, AGROAL]
The hibernate extension is registered directly with Narayana instead of the TransactionSynchronizationRegistry produced by the quarkus narayana-jta extension so I changed it to use the correct one.

I updated the reproducer (https://github.com/quarkusio/quarkus/issues/18737#issuecomment-881367249) to run with the current quarkus version (999-SNAPSHOT) and with this fix and it runs fine.

yrodiere commented on extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/internal/tsr/AgroalOrderedLastSynchronizationList.java:
```diff
@@ -0,0 +1,156 @@
+
+        // add the synchronization to the group that matches this package, if there is no matching group then add it to the
+        // catch-all group (otherSyncs)
+        String packageName = synchronization.getClass().getName();
+        SynchronizationGroup synchGroup = synchGroups.stream()
+                .filter(g -> packageName.startsWith(g.packageName)).findFirst().orElse(otherSyncs);
```
Since this is the hot path (executed for every single request involving the database), would you mind using a simple loop instead of streams/lambdas here? I seem to recall capturing lambdas in particular have a bad impact on performance.

cc @Sanne / @gsmet

mmusgrov replied:
Done

yrodiere commented on extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/internal/tsr/AgroalOrderedLastSynchronizationList.java:
```diff
@@ -0,0 +1,156 @@
+     */
+    @Override
+    public void beforeCompletion() {
+        // run each group of synchs according to the order they were added to the list
+        synchGroups.forEach(SynchronizationGroup::beforeCompletion);
+    }
+
+    @Override
+    public void afterCompletion(int status) {
+        // run each group of synchs according to the order they were added to the collection
+        synchGroups.forEach(g -> g.afterCompletion(status));
```
Same here... To be confirmed by @Sanne / @gsmet .

mmusgrov replied:
Done

yrodiere commented on independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/EventImpl.java:
```diff
@@ -291,12 +295,13 @@ void notify(T event, ObserverExceptionHandler exceptionHandler, boolean async) {
+
                             try {
                                 // NOTE - We are using standard synchronization on purpose as that seems more
                                 // fitting than interposed sync. Either way will have some use-cases that won't work.
                                 // See for instance discussions on https://github.com/eclipse-ee4j/cdi/issues/467
-                                txManager.getTransaction().registerSynchronization(sync);
+                                tsr.registerInterposedSynchronization(sync);
```
I think the comment just above is saying we shouldn't use interposed synchronization here... Can you confirm it's wrong? If so, it should probably be updated.

manovotn replied:
> @manovotn Do you recall why you changed this to a session sync (earlier it had been interposed).

No I don't, there is a comment with a link directly above this code. You'd probably need to trace it back to see why we did it. But I am pretty sure there must have been some use case for it, I don't see us just changing it on a whim.
I can try looking around tomorrow but it's quite a while ago :)

mmusgrov replied:
I thought it might be because using an interposed synchronisation meant that you couldn't guarantee when it would run and you wanted it to run before other synchronisations and using a session would enable that. But now that we can order it to run first, using an interposed synchronisation is now feasible.

manovotn replied:
This issue is also related to interposed sync and ordering of it - https://github.com/quarkusio/quarkus/issues/18737
Namely the following comment seems relevant - https://github.com/quarkusio/quarkus/issues/18737#issuecomment-911712330
The commit performing the change as such doesn't have any valuable info though.

manovotn replied:
I think one of the reasons for standard sync. was that the code in the observers is an arbitrary user code. So in a way, using tx observer is similar to writing and registering your own custom synchronization. And I was led to believe that interposed sync. was meant for container-based events rather than user code but my knowledge of JTA is shallow so I might be wrong :)

Having Arc use interposed sync. is probably OK except a case where user would also register their own synchronization in which case the ordering will be somewhat awkward.

> I thought it might be because using an interposed synchronisation meant that you couldn't guarantee when it would run

If it wasn't possible to order them back when the PR was done, that might well have been another reason for it.

Another thing probably worth noting is that Weld (or rather [WFLY integration of it](https://github.com/wildfly/wildfly/blob/main/weld/transactions/src/main/java/org/jboss/as/weld/services/bootstrap/WeldTransactionServices.java#L95)) always used standard synchronization too.

@mkouba any chance you recall why we went from interposed to standard in Arc?

mkouba replied:
> @mkouba any chance you recall why we went from interposed to standard in Arc?

I'm sorry but I have no idea :-(

manovotn replied:
> I'm sorry but I have no idea :-(

No problem.

Anyway, if a JTA expert says it should be interposed, I am not against it so long as CI ends up green :)
Just please update the comment above the code @mmusgrov

mmusgrov replied:
Thanks, I think your arguments are persuasive and I've backed out the change.

I've improved the test to cause hibernate and agroal to also register synchronizations but I hit a problem mixing `@RegisterExtension` with `@QuarkusTest` in the same test suite. I reported the problme on zulip so I will delay pushing another commit until that's resolved.

manovotn replied:
>... I've backed out the change.

Not sure I follow? I am still seeing the interposed sync in this PR

yrodiere replied:
I don't see a change to `EventImpl` anymore.

yrodiere reviewed (commented):
Thanks. Here are a few comments. Not 100% sure on the "performance" ones, but the others at least seem relevant.

mmusgrov commented on extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/customized/QuarkusJtaPlatform.java:
```diff
@@ -56,7 +69,7 @@ public Object getTransactionIdentifier(final Transaction transaction) {
 
     @Override
     public void registerSynchronization(Synchronization synchronization) {
-        this.tmSynchronizationStrategy.registerSynchronization(synchronization);
+        retrieveTransactionSynchronizationRegistry().registerInterposedSynchronization(synchronization);
```
Done

manovotn commented:
The CI failures also seem related @mmusgrov

manovotn commented:
I left comments in the Zulip thread but to sum it up for any onlooker here:
* `FruitsEndpointTest` uses different JUnit 5 extension hence the exception there, make it use `@RegisterExtension` along with `QuarkusUnitTest`
* `TSRTest` should use the same Junit 5 extension as above and to get rid of the cyclic dep, move it to the the `hibernate-orm` extension whihc already has both, hiberate and narayana dep
* Note that you'll need to add jackson dep as well to have `ObjectMapper` bean accessible

Hope this helps

mmusgrov commented:
> I left comments in the Zulip thread but to sum it up for any onlooker here:
> 
> * `FruitsEndpointTest` uses different JUnit 5 extension hence the exception there, make it use `@RegisterExtension` along with `QuarkusUnitTest`
> * `TSRTest` should use the same Junit 5 extension as above and to get rid of the cyclic dep, move it to the the `hibernate-orm` extension whihc already has both, hiberate and narayana dep
> * Note that you'll need to add jackson dep as well to have `ObjectMapper` bean accessible
> 
> Hope this helps

Thanks @manovotn , it helped a lot.

The failure is in the panache extension and looks to be unrelated to my change and I don't know how to access the surefire logs so could not check (/home/runner/work/quarkus/quarkus/extensions/panache/hibernate-orm-rest-data-panache/deployment/target/surefire-reports).

mmusgrov commented:
> @mmusgrov you can download build logs from GH actions and inspect those (or view raw logs in browser). Open the CI run that's failing and in the upper right corner click options (the cogwheel) and choose whichever you prefer. It's not exactly the surefire report but it at least tells you something.
> 
> Alternatively, you can run local build (`mvn clean install -Dquickly`) and then tests only for the affected module.
> 
> Looking just at the raw logs, I can see failures in `PanacheEntityResourcePostMethodTest`.
> 

Thanks @manovotn , I will investigate and update my PR when tracked down the cause.

mmusgrov commented:
My latest push should fix the test failures, I also git rebased but I don't understand why it has included `@michalvavrik Move form auth and basic realm to runtime 2651ca1`.

mmusgrov committed:
Ordering of interposed synchronizations

mmusgrov commented:
The only failure seems to be related to the CI build/test environment:
```
Error:  Tests run: 1, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 18.13 s <<< FAILURE! -- in io.quarkus.gradle.CustomManifestArgumentsTest
Error:  io.quarkus.gradle.CustomManifestArgumentsTest.shouldContainsSpecificManifestProperty -- Time elapsed: 18.12 s <<< FAILURE!
```
and the failing test is https://github.com/quarkusio/quarkus/blob/main/integration-tests/gradle/src/test/java/io/quarkus/gradle/CustomManifestArgumentsTest.java and is unreleated to any extension.

So this PR is ready to be reviewed again. @manovotn I know you've had some good input so far, is this something you could take a look at or should I tag one of the other reviewers? FYI it was issue [#36877](https://github.com/quarkusio/quarkus/issues/36877) that prompted me to progress this PR.

manovotn commented:
@mmusgrov I've marked all of us that commented here previously. I'll try to review today/tomorrow.

yrodiere reviewed (commented):
Thanks. I don't see a problem with this, though I can't claim I fully grasp the implications :)
I'll let @manovotn review as well.

manovotn reviewed (approved):
I can't claim I understand all the parts but the overall idea and impl looks good to me.

mmusgrov commented:
@scottmarlow Yoann Rodière and Matej Novotny both approve of the PR but both say they can't claim to understand all the parts but the overall idea and impl looks good to them. I know that you have some experience in this area, is this something you could take a look at or should I try one of the other experts in this area?

scottmarlow commented on extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/internal/tsr/AgroalOrderedLastSynchronizationList.java:
```diff
@@ -0,0 +1,192 @@
+     */
+    private enum ExecutionStatus {
+        PENDING, // the synchronization has not been called yet
+        RUNNING, // the synchronization is execution
```
minor typo in comment, `execution` should be `executing` I think.

scottmarlow commented on extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/internal/tsr/AgroalOrderedLastSynchronizationList.java:
```diff
@@ -0,0 +1,192 @@
+     */
+    private class SynchronizationGroup implements Synchronization {
+        String packagePrefix; // Synchronizations with this package prefix belong to this group
+        List<Synchronization> synchs; // the Synchronizations in the group
```
should `synchs` be final?

mmusgrov replied:
Agreed

scottmarlow commented on extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/internal/tsr/AgroalOrderedLastSynchronizationList.java:
```diff
@@ -0,0 +1,192 @@
+    private class SynchronizationGroup implements Synchronization {
+        String packagePrefix; // Synchronizations with this package prefix belong to this group
+        List<Synchronization> synchs; // the Synchronizations in the group
+        ExecutionStatus status; // track the status to decide when it's too late to allow more registrations
```
Should `status` be volatile so it doesn't get placed in a CPU register (or whatever thread memory is considered to be) or read from a CPU register?

mmusgrov replied:
Agreed

scottmarlow commented on extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/internal/tsr/TransactionSynchronizationRegistryWrapper.java:
```diff
@@ -0,0 +1,92 @@
+ * <p>
+ *
+ * Synchronizations are placed into groups according to their package name and the groups are ordered which means
+ * that all hibernate synchronizations run before Agroal ones and all other synchs run before the hibernate ones.
```
:+1:

scottmarlow commented on extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/internal/tsr/AgroalOrderedLastSynchronizationList.java:
```diff
@@ -0,0 +1,192 @@
+     * @throws IllegalStateException if the transaction is in the wrong state:
+     *         <ol>
+     *         <li>the transaction has already prepared;
+     *         <li>the transaction is marked rollback only
```
Are we throwing IllegalStateException because we want to fail-fast the marked for rollback only transaction?  If yes, the comment and the code below is fine.  If that isn't the intention then why are we throwing a failure when the transaction cannot succeed so not much else to do?

mmusgrov replied:
It's a good question. I am throwing the exception here based on what we do in Narayana, see [1] and [2], where we throw an exception if it's marked for rollback. So, although I didn't write that Arjuna code, I would say it was done in order to fail fast.

[1] https://github.com/jbosstm/narayana/blob/7.0.0.Final/ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/transaction/arjunacore/TransactionImple.java#L371
[2] https://github.com/jbosstm/narayana/blob/7.0.0.Final/ArjunaCore/arjuna/classes/com/arjuna/ats/arjuna/coordinator/TwoPhaseCoordinator.java#L116

mmusgrov commented:
The build failure is because the `formatter-maven-plugin` reports that one of the files I added needs a reformat - sorry about that. I'll push an update today.

mmusgrov commented:
FYI The PR isn't ready for merging yet - it is using the TSR resource map incorrectly so I need to push an update to fix it and an extra test.

mmusgrov commented:
Thanks @yrodiere for marking it as a draft.
I added the tests and pushed the update last night but the `CI Sanity Check` is not making progress.

manovotn commented:
> Thanks @yrodiere for marking it as a draft. I added the tests and pushed the update last night but the `CI Sanity Check` is not making progress.

Seems like the PR needed approval to run which I've given it.
I've also moved it back to ready for review since you pushed and the CI doesn't run all the tests for drafts either.

mmusgrov commented:
@scottmarlow @manovotn

I made two further changes (one functional and one that adds more testing) to the PR since you last approved it:

- I moved the deferredThrowable (https://github.com/quarkusio/quarkus/pull/34160/commits/7b02affa17cf2c1a1d6e8d6dff18d0fdc9b737f2#diff-eb042caa54555177c75fa37d6a25c9718840b45ee89997cf60fea2d6e5e15331R24) from the inner class to the outer on since it needs to be thrown after processing *all* of the synchronizations.
- I added a test showing that exceptions thrown by a synchronization are reported correctly (https://github.com/quarkusio/quarkus/pull/34160/commits/7b02affa17cf2c1a1d6e8d6dff18d0fdc9b737f2#diff-0e882ca4edb1152dd8c9035a45d829f39d76d083b757ddcf535986f5b3a7a822R114);
- I added tests to QuarkusTransactionTest.java which validate Synchronizations when multiple threads and transactions are active (https://github.com/quarkusio/quarkus/pull/34160/commits/7b02affa17cf2c1a1d6e8d6dff18d0fdc9b737f2#diff-187f40d06a6bbfbb79aecf3bd9b0d0c241f464070beec185d56b9c746e511519)

Are your approvals still valid?

scottmarlow commented on extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/internal/tsr/AgroalOrderedLastSynchronizationList.java:
```diff
@@ -0,0 +1,190 @@
+     */
+    @Override
+    public void beforeCompletion() {
+        // run each group of synchs according to the order they were added to the list
```
I expect that both beforeCompletion + afterCompletion method calls will invoke `Hibernate` second to last and `Agroal` last.  I'm not really sure though if the above code is always adding `Hibernate` second to last in the list and `Agroal` last in the list.  Did I miss something that ensures that Hibernate/Agroal are added to the list last?

mmusgrov replied:
The synchronization groups are maintained in an ordered list [1]. There are three such groups and the members are based on the package names [2] and [3] of the synchronization being registered. The order of the package names [3] is '{ "", "org.hibernate", "io.agroal.narayana" };` so the second group of synchronizations is any synchronization with package prefix "org.hibernate" and the last group contains synchronizations with package prefix "io.agroal.narayana".

The code you have highlighted iterates through each group in order and calls `beforeCompletion()` on the group which in turn calls `beforeCompletion()` on the actual registered synchronizations [5].

[1] https://github.com/mmusgrov/quarkus/blob/18737-synch-order/extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/internal/tsr/AgroalOrderedLastSynchronizationList.java#L21
[2] https://github.com/mmusgrov/quarkus/blob/18737-synch-order/extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/internal/tsr/AgroalOrderedLastSynchronizationList.java#L115
[3] https://github.com/mmusgrov/quarkus/blob/18737-synch-order/extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/internal/tsr/AgroalOrderedLastSynchronizationList.java#L19
[4] https://github.com/mmusgrov/quarkus/blob/18737-synch-order/extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/internal/tsr/AgroalOrderedLastSynchronizationList.java#L168
[5] https://github.com/mmusgrov/quarkus/blob/18737-synch-order/extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/internal/tsr/AgroalOrderedLastSynchronizationList.java#L71

mmusgrov commented:
Thanks everyone for helping out with the reviews. @yrodiere I think we have everything in place now.

yrodiere commented:
> Thanks everyone for helping out with the reviews. @yrodiere I think we have everything in place now.

It seem there's a conflict now, because I managed to merge the upgrade to Hibernate ORM 6.4 before you got all your approvals... Sorry :X

Can you rebase, please?

yrodiere commented:
Thanks. Let's wait for CI and then we can merge.

yrodiere commented:
Merged, thanks everyone!
